### PR TITLE
Implement chat pages

### PR DESCRIPTION
**Pull request by Dune-jr**
- Added some console-like chat `scrolling` system with pageup/pagedown. There are no scrollbars for now, but it might be better to make it a scrollable region instead of pages.

screenshot_2019-08-19_21-35-50
screenshot_2019-08-19_21-35-57
screenshot_2019-08-19_21-36-09

- Fix some issue with chat rendering being aborted on the first message (didn't check the bug really did happen but that `break` didn't look good)

### CODE REVIEW

File: src/game/client/components/chat.cpp [hunk 1]
```diff
@@ -196,7 +196,30 @@ void CChat::ClearChatBuffer()
...
+			else if(Event.m_Key == KEY_PAGEDOWN)
+			{
+				--m_BacklogPage;
+				if(m_BacklogPage < 0)
+					m_BacklogPage = 0;
+			}
+		}
+		return true;
+	}
+	if(m_Mode == CHAT_NONE)
```
Comment (oy) [inline src/game/client/components/chat.cpp]: This check should probably be up, so you don't trigger it when the chat is closed. Guess ingame binds for PageUp/Down don't work either?

File: src/game/client/components/chat.h [hunk 2]
```diff
@@ -52,6 +53,7 @@ class CChat : public CComponent
 	int m_WhisperTarget;
 	int m_LastWhisperFrom;
 	bool m_Show;
+	int m_BacklogPage;
```
Comment (oy) [inline src/game/client/components/chat.h]: This should be initialized somewhere

Comment (Dune-jr) [reply src/game/client/components/chat.h]: Good catch, thanks

**Comment by Dune-jr** [comment]
No computer available but on second thought, I believe 72ef4a0 breaks the page scrolling when show_chat is enabled. I think it is best to revert it.

**Comment by oy** [comment]
True. Yeah, then revert the last one.

File: src/game/client/components/chat.cpp [hunk 3]
```diff
@@ -196,8 +197,31 @@ void CChat::ClearChatBuffer()
...
+					m_BacklogPage = MAX_CHAT_PAGES-1;
+			}
+			else if(Event.m_Key == KEY_PAGEDOWN)
+			{
+				--m_BacklogPage;
+				if(m_BacklogPage < 0)
+					m_BacklogPage = 0;
+			}
+		}
+		return true;
```
Comment (oy) [inline src/game/client/components/chat.cpp]: Could check if `m_Mode == CHAT_NONE` is true and then return false, then ingame binds should work.

Comment (Dune-jr) [reply src/game/client/components/chat.cpp]: Done

Comment (Dune-jr) [reply src/game/client/components/chat.cpp]: @oy Hmm, so I think what you are thinking of is 
```cpp
	// chat history scrolling
	if(m_Show && Event.m_Flags&IInput::FLAG_PRESS && (Event.m_Key == KEY_PAGEUP || Event.m_Key == KEY_PAGEDOWN))
	{
		if(Event.m_Key == KEY_PAGEUP)
		{
			++m_BacklogPage;
			if(m_BacklogPage >= MAX_CHAT_PAGES) // will be further capped during rendering
				m_BacklogPage = MAX_CHAT_PAGES-1;
		}
		if(Event.m_Key == KEY_PAGEDOWN)
		{
			--m_BacklogPage;
			if(m_BacklogPage < 0)
				m_BacklogPage = 0;
		}
		return (m_Mode != CHAT_NONE); // trigger binds if not chatting
	}
```

I am not sure if it's such a good idea to trigger pageup/pagedown binds when they are already hardcoded and used for the chat pages though?

**Comment by Dune-jr** [comment]
Following yesterday's discussion, I would like to define a setting for `previous page` / `next page`, that could be used for other GUI elements such as a 64p scoreboard.

While this would work just fine when `(m_Mode == CHAT_NONE)`, when it is not the case the inputs will be captured by the chat and the binds would not be triggered.

What would be a good solution here? It seems like there needs to be something in `Chat::OnInput` either way :/

**Comment by oy** [comment]
Yeah, tricky one. Have to think about a good solution :)

**Comment by Dune-jr** [comment]
I suggest to use https://github.com/teeworlds/teeworlds/pull/2195/commits/1fbbd6518535cead72996e9c3a0fee9bf637ff7c for now. We can always implement a page system later when it comes in useful for other UI elements. Being able to customize pageup/pagedown doesn't seem very important to me.

I will have to rebind my `bind pagedown disconnect` though :p

**Pull request merged**